# TM-MSAligner
## A Multi-Objective Multiple Sequence Alignment Tool for Transmembrane Proteins

TM-MSAligner is a multiobjective software tool to align multiple transmembrane protein sequences
TM-MSAligner is based in the [M2Align](https://github.com/KhaosResearch/M2Align) software,
which use the reference multiobjective metaheuristics NSGAII and is extended with an encoding 
for representing MSA solutions and its corresponding mutation and crossover operators, and it is able of reducing the computing time by exploiting the computing capabilities of common multi-core CPU computers

Impact of the Software TM-MSAligner applied to BAlibase Ref7

## Requirements
To use TM-MSAligner the following software packages are required:
* [Java SE Development Kit 8](http://www.oracle.com/technetwork/java/javase/downloads/jdk8-downloads-2133151.html?ssSourceSiteId=otnes)
* [Apache Maven](https://maven.apache.org/)
* [Git](https://git-scm.com/)

## Downloading and compiling
To download TM-MSAligner just clone the Git repository hosted in GitHub:
```
git clone https://github.com/cristianzambrano/TM-MSAligner.git
```
Once cloned, you can compile the software and generate a jar file with the following command:

```
mvn package
```

This sentence will generate a directory called `target` which will contain a file called `tm-msaligner-1.0-SNAPSHOT-jar-with-dependencies.jar`

## Runing TM-MSAligner

To execute the TM-MSAligner algorithm to align a custom dataset of sequences, just run this command:

``` java
java -cp target/tm-msaligner-1.0-SNAPSHOT-jar-with-dependencies.jar org.uma.khaos.tm_msaligner.TM_M2Align_Runner dataDirectory problemName NumberOfEvaluations PopulationSize NumberOfCores typeObserver 
```

data/custom_tests/ msl 25000 100 1 0

* dataDirectory: The path of the data directory. This directory must contain a subfolder named as the problemName,
  the TMP sequences with the extension .3line (generated by ..) and the Pre-Computed alignments (in FASTA Format) performed by state-of-the-art MSA software used to generate the Initial population of the algorithm.
  The algorithm search files with extension `.fasta`. See details how to generate them in the next section.
* problemName: Name of the custom TMP Sequences dataset
* NumberOfEvaluations PopulationSize NumberOfCores
* NumberOfEvaluations: Number of the Maximun Evaluations of the algorithm.
* PopulationSize: Size of the population of the algorithm
* NumberOfCores: Number of cores to use for the execution of the Algorithm, greater than 1 enable the Parallel features.
* typeObserver: Type of the Observer added to the algorithm. 0: None, 1: FitnessWriteFileObserver, 2: FitnessPlotObserver and 3: FrontPlotTM_MSAObserve

To execute the algorithm to solve a problem in Balibase ref7 dataset, just run this command:

``` java
java -cp target/m2align-1.0-SNAPSHOT-jar-with-dependencies.jar TM_M2AlignBAliBASETest_Runner balibaseInstanceName dataDirectory NumberOfEvaluations PopulationSize NumberOfCores
```
* balibaseInstanceName: the BAliBASE instance name, for instance `photo`.
* typeObserver: Type of the Observer added to the algorithm. 0: None, 1: FitnessWriteFileObserver, 2: FitnessPlotObserver and 3: FrontPlotTM_MSAObserve

For solving BAliBASE problems, TM-MSAligner searches: the Sequence File in FASTA format (.fasta) and the pre-computed alignments, as follows:

* benchmarkPath: Directory with the original sequence without align, join the path as follows 'data/benchmarks/ref7/' + `balibaseInstanceName` + '/'
* preComputedMSAPath: Directory with the PreAlignments, oin the path as follows 'data/precomputed_solutions/ref7/' + `balibaseInstanceName` +'/'
* dataFile: The filename of the original TMP sequences of BAliBASE instance, 'benchmarkPath' + `balibaseInstanceName` + '_predicted_topologies.3line'

* String outputFolder = "data/pruebas/ref7/" + refName + "/test" + System.currentTimeMillis() +"/" ;


## Results

Visualizations of Pareto Front Solutions and Alignments

Ilustrating  Solution X:
![Frente de Pareto](images/HtmlViewer1.png)

Ilustrating  Solution X:
![Frente de Pareto](images/HtmlViewer2.png)


## Commands to precompute alignments (needed for generating the initial population)
Assuming that the FASTA file containing the sequences to align is named "FASTA_FILE", the sintax to run the different tools are:
```
clustalw2 -infile=FASTA_FILE -outfile=FASTA_FILE_clu" -output=FASTA -ALIGN -QUIET -OUTORDER=input
mafft --auto --inputorder --quiet FASTA_FILE > FASTA_FILE_mafft
t_coffee FASTA_FILE -output fasta
muscle -in FASTA_FILE -fastaout FASTA_FILE_muscle" -quiet
java -jar /usr/local/bin/retalign.jar -out FASTA_FILE_retalign" FASTA_FILE
kalign -c input -f fasta -q -i FASTA_FILE -o FASTA_FILE_kalign
probcons FASTA_FILE > FASTA_FILE_probcons
fsa --refinement 100 FASTA_FILE > FASTA_FILE_fsa
