# TM-MSAligner
## A Multi-Objective Multiple Sequence Alignment Tool for Transmembrane Proteins

TM-MSAligner is a multiobjective software tool to align multiple transmembrane protein sequences (TMPs)
TM-MSAligner is based in the [M2Align](https://github.com/KhaosResearch/M2Align) software,
which use the reference multiobjective metaheuristics NSGAII and is extended with an encoding 
for representing MSA solutions and its corresponding mutation and crossover operators, and it is able of reducing the computing time by exploiting the computing capabilities of common multi-core CPU computers

Impact of the Software TM-MSAligner applied to BAlibase Ref7

## Requirements
To use TM-MSAligner the following software packages are required:
* [Java SE Development Kit 8](http://www.oracle.com/technetwork/java/javase/downloads/jdk8-downloads-2133151.html?ssSourceSiteId=otnes)
* [Apache Maven](https://maven.apache.org/)
* [Git](https://git-scm.com/)

## Downloading and compiling
To download TM-MSAligner just clone the Git repository hosted in GitHub:
```
git clone https://github.com/cristianzambrano/TM-MSAligner.git
```
Once cloned, you can compile the software and generate a jar file with the following command:

```
mvn package
```

This sentence will generate a directory called `target` which will contain a file called `tm-msaligner-1.0-SNAPSHOT-jar-with-dependencies.jar`

## Generatye the Topology Prediction using TM-MSAligner



## Runing TM-MSAligner

### Running TM-MSAligner to align a custom TMPs sequences dataset
To execute the TM-MSAligner algorithm to align a custom dataset of sequences, just run this command:

``` java
java -cp target/tm-msaligner-1.0-SNAPSHOT-jar-with-dependencies.jar org.uma.khaos.tm_msaligner.TM_M2Align_Runner dataDirectory problemName NumberOfEvaluations PopulationSize NumberOfCores typeObserver 
```

* dataDirectory: The path of the data directory. This directory must contain two components. First, the file with the TMPs sequences and the topology prediction information, called `problemName_predicted_topologies.3line` (See how generate this file in above section).
And second, the Pre-Computed alignments files in FASTA format (`.fasta` extension) performed by state-of-the-art of MSA software, See details how to generate them in the section below.
These pre-computed alignments are used to generate the Initial population of the algorithm. The algorithm search files with extension `.fasta` and add to the list.
* problemName: Name of the custom TMP Sequences dataset.
* NumberOfEvaluations: Number of the maximun evaluations of the algorithm.
* PopulationSize: Size of the population of the algorithm.
* NumberOfCores: Number of cores to use for the execution of the Algorithm, greater than 1 enable the Parallel features.
* observerType: Type of the Observer added to the algorithm. TM-MSAligner includes three different observers to plot the performance of the algorithm, use the following values:  0 -> None, 1 -> FitnessWriteFileObserver, 2 -> FitnessPlotObserver and 3 -> FrontPlotTM_MSAObserve. See details of each observer in the section below.

An example to align a custom TMPs dataset called `msl` stored in the folder `data/custom_tests`, with 25000 evaluations of the algorithm, a population size of 100, using 4 cores and plotting the ParetoFront every each 500 evaluations,
you have to run this command:

``` java
java -cp target/tm-msaligner-1.0-SNAPSHOT-jar-with-dependencies.jar org.uma.khaos.tm_msaligner.TM_M2Align_Runner data/custom_tests/ msl 25000 100 4 3
```

The results are saved in the folder called as: `dataDirectory`/results_`currentSystemTimeMilliseconds`/. See more details about the results generated by our software in the next section.

### Runing TM-MSAligner to align a TMPs instance from BAliBASE benchmark
To evaluate the performance of our software, we have used some multiple sequence alignment (MSA) benchmarkings, one of them, is the reference 7 of BAliBASE (Benchmark Alignment dataBASE), 
this dataset is made of 435 alpha-helical TMPs classified into 8 distinct families named: 7tm, msl, ion, acr, nat, dtd, photo and ptga.

We have added the BAliBASE-ref7 dataset in our [data](https://github.com/cristianzambrano/TM-MSAligner/tree/main/data) folder. 
In this folder we have added the pre-computed alignments performed by state-of-the-art of MSA software: ClustalW, KAlign, TM-TCoffee, Muscle, Mafft, T-Coffee.
Furthermore, we have added the `refName`_predicted_topologies.3line for each BAliBASE-ref7 instance, saved in the data/benchmarks/ref7 folder.

To execute the algorithm to solve a problem of the BAliBASE TMP's benchmark (ref7), just run this command:

``` java
java -cp target/tm-msaligner-1.0-SNAPSHOT-jar-with-dependencies.jar TM_M2AlignBAliBASETest_Runner balibaseInstanceName dataDirectory NumberOfEvaluations PopulationSize NumberOfCores
```
* balibaseInstanceName: the BAliBASE ref7 instance name.
* observerType: Type of the Observer added to the algorithm. TM-MSAligner includes three different observers to plot the performance of the algorithm, use the following values:  0 -> None, 1 -> FitnessWriteFileObserver, 2 -> FitnessPlotObserver and 3 -> FrontPlotTM_MSAObserve. See details of each observer in the next section.

Solving BAliBASE-ref7 benchmarking, TM-MSAligner considers the following pathfolders:

* benchmarkPath: Directory with the original BAliBASE-ref7 sequences without align and the predicted_topologies.3line file, the path is joined as follows 'data/benchmarks/ref7/' + `balibaseInstanceName` + '/'
* dataFile: The filename of the original TMP sequences of BAliBASE instance, 'benchmarkPath' + `balibaseInstanceName` + '_predicted_topologies.3line'
* preComputedMSAPath: Directory with the PreAlignments, TM-MSAligner reads all `.fasta` extension files stored in this folder, the path is joined as follows 'data/precomputed_solutions/ref7/' + `balibaseInstanceName` +'/'

The results are saved in the folder called as: data/tests/ref7/`balibaseInstanceName`/results_`currentSystemTimeMilliseconds`/. See more details about the results generated by our software in the next section.

## Results

Visualizations of Pareto Front Solutions and Alignments

Ilustrating  Solution X:
![Frente de Pareto](images/HtmlViewer1.png)

Ilustrating  Solution X:
![Frente de Pareto](images/HtmlViewer2.png)


## Commands to precompute alignments (needed for generating the initial population)
Assuming that the FASTA file containing the sequences to align is named "FASTA_FILE", the sintax to run the different tools are:
```
clustalw2 -infile=FASTA_FILE -outfile=FASTA_FILE_clu" -output=FASTA -ALIGN -QUIET -OUTORDER=input
mafft --auto --inputorder --quiet FASTA_FILE > FASTA_FILE_mafft
t_coffee FASTA_FILE -output fasta
muscle -in FASTA_FILE -fastaout FASTA_FILE_muscle" -quiet
java -jar /usr/local/bin/retalign.jar -out FASTA_FILE_retalign" FASTA_FILE
kalign -c input -f fasta -q -i FASTA_FILE -o FASTA_FILE_kalign
probcons FASTA_FILE > FASTA_FILE_probcons
fsa --refinement 100 FASTA_FILE > FASTA_FILE_fsa
